parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestMemoryManager.java

```diff
@@ -44,89 +44,146 @@
-    Assert.assertEquals("memory pool size is incorrect.", expectPoolSize,
-        parquetOutputFormat.getMemoryManager().getTotalMemoryPool());
+  @Test
+  public void testReallocationCallback() throws Exception {
+    // validate assumptions
+    long poolSize = ParquetOutputFormat.getMemoryManager().getTotalMemoryPool();
+    long rowGroupSize = poolSize / 2;
+    conf.setLong(ParquetOutputFormat.BLOCK_SIZE, rowGroupSize);
+
+    Assert.assertTrue("Pool should hold 2 full row groups",
+        (2 * rowGroupSize) <= poolSize);
+    Assert.assertTrue("Pool should not hold 3 full row groups",
+        poolSize < (3 * rowGroupSize));
+
+    Runnable callback = new Runnable() {
+      @Override
+      public void run() {
+        counter++;
+      }
+    };
 
-    //Verify Callback mechanism
-    Assert.assertEquals("counter calculated by callback is incorrect.", 1, counter);
-    Assert.assertEquals("CallBack is duplicated.", 1, parquetOutputFormat.getMemoryManager()
-        .getScaleCallBacks().size());
-  }
+    // first-time registration should succeed
+    ParquetOutputFormat.getMemoryManager()
+        .registerScaleCallBack("increment-test-counter", callback);
 
-  private RecordWriter createWriter(int index) throws Exception{
-    Path file = new Path("target/test/", "parquet" + index);
-    parquetOutputFormat = new ParquetOutputFormat(new GroupWriteSupport());
-    RecordWriter writer = parquetOutputFormat.getRecordWriter(conf, file, codec);
     try {
-      parquetOutputFormat.getMemoryManager().registerScaleCallBack("increment-test-counter",
-          new Runnable() {
-            @Override
-            public void run() {
-              counter++;
-            }
-          });
-      if (!firstRegister) {
-        Assert.fail("Duplicated registering callback should throw duplicates exception.");
-      }
-      firstRegister = false;
+      ParquetOutputFormat.getMemoryManager()
+          .registerScaleCallBack("increment-test-counter", callback);
+      Assert.fail("Duplicated registering callback should throw duplicates exception.");
     } catch (IllegalArgumentException e) {
-      if (firstRegister) {
-        Assert.fail("Registering the same callback first time should succeed.");
-      }
+      // expected
     }
 
+    // hit the limit once and clean up
+    RecordWriter writer1 = createWriter(1);
+    RecordWriter writer2 = createWriter(2);
+    RecordWriter writer3 = createWriter(3);
+    writer1.close(null);
+    writer2.close(null);
+    writer3.close(null);
+
+    //Verify Callback mechanism
+    Assert.assertEquals("Allocations should be adjusted once", 1, counter);
+    Assert.assertEquals("Should not allow duplicate callbacks",
+        1, ParquetOutputFormat.getMemoryManager().getScaleCallBacks().size());
+  }
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private RecordWriter createWriter(int index) throws Exception {
+    File file = temp.newFile(String.valueOf(index) + ".parquet");
```

Comment:
Are we deleting this file after it's being used? Mabye a file.deleteOnExit() here would help to avoid leaving temporary files after tests.